File: android/src/main/kotlin/twilio/flutter/twilio_programmable_chat/methods/UsersMethods.kt
package twilio.flutter.twilio_programmable_chat.methods

import com.twilio.chat.CallbackListener
import com.twilio.chat.ErrorInfo
import com.twilio.chat.Paginator
import com.twilio.chat.User
import com.twilio.chat.UserDescriptor
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import twilio.flutter.twilio_programmable_chat.Mapper
import twilio.flutter.twilio_programmable_chat.PaginatorManager
import twilio.flutter.twilio_programmable_chat.TwilioProgrammableChatPlugin

object UsersMethods {
    fun getChannelUserDescriptors(call: MethodCall, result: MethodChannel.Result) {
        val channelSid = call.argument<String>("channelSid")
                ?: return result.error("ERROR", "Missing 'channelSid'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.users?.getChannelUserDescriptors(channelSid, object : CallbackListener<Paginator<UserDescriptor>>() {
            override fun onSuccess(paginator: Paginator<UserDescriptor>) {
                TwilioProgrammableChatPlugin.debug("UsersMethods.getChannelUserDescriptors => onSuccess")
                val pageId = PaginatorManager.setPaginator(paginator)
                result.success(Mapper.paginatorToMap(pageId, paginator, "userDescriptor"))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("UsersMethods.getChannelUserDescriptors => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getUserDescriptor(call: MethodCall, result: MethodChannel.Result) {
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.users?.getUserDescriptor(identity, object : CallbackListener<UserDescriptor>() {
            override fun onSuccess(userDescriptor: UserDescriptor) {
                TwilioProgrammableChatPlugin.debug("UsersMethods.getUserDescriptor => onSuccess")
                result.success(Mapper.userDescriptorToMap(userDescriptor))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("UsersMethods.getUserDescriptor => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getAndSubscribeUser(call: MethodCall, result: MethodChannel.Result) {
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        TwilioProgrammableChatPlugin.chatListener.chatClient?.users?.getAndSubscribeUser(identity, object : CallbackListener<User>() {
            override fun onSuccess(user: User) {
                TwilioProgrammableChatPlugin.debug("UsersMethods.getAndSubscribeUser => onSuccess")
                result.success(Mapper.userToMap(user))
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioProgrammableChatPlugin.debug("UsersMethods.getAndSubscribeUser => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }
}

File: ios/Classes/Methods/UsersMethods.swift
import Flutter
import Foundation
import TwilioChatClient

public class UsersMethods {
    public static func getChannelUserDescriptors(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let channelSid = arguments["channelSid"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'channelSid' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.channelsList()?.channel(withSidOrUniqueName: channelSid, completion: { (result: TCHResult, channel: TCHChannel?) in
            if result.isSuccessful(), let channelRes = channel as TCHChannel? {
                SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.users()?.userDescriptors(for: channelRes, completion: { (result: TCHResult, userDescriptors: TCHUserDescriptorPaginator?) in
                    if result.isSuccessful(), let descriptors = userDescriptors as TCHUserDescriptorPaginator? {
                        SwiftTwilioProgrammableChatPlugin.debug("UsersMethods.getChannelUserDescriptors => onSuccess")
                        let pageId = PaginatorManager.setPaginator(descriptors)
                        flutterResult(Mapper.userDescriptorsToDict(pageId, descriptors))
                    } else {
                        SwiftTwilioProgrammableChatPlugin.debug("UsersMethods.getChannelUserDescriptors => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error retrieving user descriptors for channel: \(channelSid)", details: nil))
                    }
                })
            } else {
                SwiftTwilioProgrammableChatPlugin.debug("UsersMethods.getChannelUserDescriptors => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving channel: \(channelSid)", details: nil))
            }
        })
    }

    public static func getUserDescriptor(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let identity = arguments["identity"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'identity' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.users()?.userDescriptor(withIdentity: identity, completion: { (result: TCHResult, user: TCHUserDescriptor?) in
            if result.isSuccessful() {
                if let userDesc = user as TCHUserDescriptor? {
                    SwiftTwilioProgrammableChatPlugin.debug("UsersMethods.getUserDescriptor => onSuccess")
                    flutterResult(Mapper.userDescriptorToDict(userDesc))
                } else {
                    flutterResult(FlutterError(code: "ERROR", message: "Could not resolve userDescriptor for identity: \(identity)", details: nil))
                }
            } else {
                SwiftTwilioProgrammableChatPlugin.debug("UsersMethods.getUserDescriptor => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving userDescriptor for identity: \(identity)", details: nil))
            }
        })
    }

    public static func getAndSubscribeUser(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let identity = arguments["identity"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'identity' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioProgrammableChatPlugin.chatListener?.chatClient?.users()?.userDescriptor(withIdentity: identity, completion: { (result: TCHResult, user: TCHUserDescriptor?) in
            if result.isSuccessful(), let user = user {
                user.subscribe(completion: { (result: TCHResult, user: TCHUser?) in
                    if result.isSuccessful() {
                        flutterResult(Mapper.userToDict(user))
                    } else {
                        flutterResult(FlutterError(code: "ERROR", message: "Error subscribing to user with identity: \(identity)", details: nil))
                    }
                })
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving user descriptor for identity: \(identity)", details: nil))
            }
        })
    }
}
